plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/cwc/inline/InlineChatPopupFactory.kt

```diff
@@ -0,0 +1,170 @@
+import com.intellij.openapi.application.ReadAction
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.editor.VisualPosition
+import com.intellij.openapi.ui.popup.IconButton
+import com.intellij.openapi.ui.popup.JBPopup
+import com.intellij.openapi.ui.popup.JBPopupFactory
+import com.intellij.openapi.util.TextRange
+import com.intellij.ui.IdeBorderFactory
+import com.intellij.ui.awt.RelativePoint
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+import software.aws.toolkits.jetbrains.services.codewhisperer.util.CodeWhispererColorUtil.POPUP_BUTTON_BORDER
+import software.aws.toolkits.resources.AmazonQBundle.message
+import java.awt.Point
+
+class InlineChatPopupFactory(
+    private val submitHandler: suspend (String, String, Int, Editor) -> String,
+    private val acceptHandler: () -> Unit,
+    private val rejectHandler: () -> Unit,
+    private val cancelHandler: () -> Unit,
+    private val popupBufferHeight: Int = 150,
+) : Disposable {
+
+    private fun getSelectedText(editor: Editor): String = ReadAction.compute<String, Throwable> {
+        val selectionStartOffset = editor.selectionModel.selectionStart
+        val selectionEndOffset = editor.selectionModel.selectionEnd
+        if (selectionEndOffset > selectionStartOffset) {
+            val selectionLineStart = editor.document.getLineStartOffset(editor.document.getLineNumber(selectionStartOffset))
+            val selectionLineEnd = editor.document.getLineEndOffset(editor.document.getLineNumber(selectionEndOffset))
+            editor.document.getText(TextRange(selectionLineStart, selectionLineEnd))
+        } else {
+            ""
+        }
+    }
+
+    private fun getSelectionStartLine(editor: Editor): Int = ReadAction.compute<Int, Throwable> {
+        editor.document.getLineNumber(editor.selectionModel.selectionStart)
+    }
+
+    fun createPopup(editor: Editor, scope: CoroutineScope): JBPopup {
+        val popupPanel = InlineChatPopupPanel(this).apply {
+            border = IdeBorderFactory.createRoundedBorder(10).apply {
+                setColor(POPUP_BUTTON_BORDER)
+            }
+
+            val submitListener: () -> Unit = {
+                submitButton.isEnabled = false
+                cancelButton.isEnabled = false
+                textField.isEnabled = false
+                val prompt = textField.text
+                if (prompt.isNotBlank()) {
+                    setLabel(message("amazonqInlineChat.popup.generating"))
+                    revalidate()
+
+                    scope.launch {
+                        val selectedCode = getSelectedText(editor)
+                        val selectedLineStart = getSelectionStartLine(editor)
+                        var errorMessage = ""
+                        runBlocking {
+                            errorMessage = submitHandler(prompt, selectedCode, selectedLineStart, editor)
+                        }
```

Comment:
not need `var` here? `val  errorMsg = runBlocking {
                            errorMessage = submitHandler(prompt, selectedCode, selectedLineStart, editor)
                        }`

Reply:
it's used to set other error messages too, not just the api ones